### ISSUE TITLE
copy styles between Word documents - can we have add_style(Style object)?

### ISSUE THREAD

**Issue by mieczyslaw** [python-openxml/python-docx#508]
I am trying to copy some styles (various types: paragraph, character, list and table) between Word documents. I am attaching the fragments of the code below. I can read styles from the source document with:

source_document = Document(source_file)
source_styles = source_document.styles

then I am opening the target document and I would like to copy some of the styles there (I can filter myself by style name property):

target_document = Document(full_path)
target_styles = target_document.styles

for s_name in ALLOWED_STYLES:
   if s_name in target_styles:
      target_styles[s_name].delete()
      source_style = source_styles[s_name]
      style = target_styles.add_style(s_name, source_style.type)
      style = source_styles[s_name] # does not work as expected, the style definition is not copied
      #target_document.styles[s_name] = copy.copy(source_styles[s_name]) # this is not allowed at all
target_document.save(full_path)

If I run the code above, the styles are added but they don't have the correct properties. It looks that Style objects cannot be assigned. One solution would be to add the second version of add_style(), which would accept the Style object. Alternatively I know that you can set the properties of the style, but I don't know how to iterate over all properties of the source styles to copy them between Style objects. 

Thanks for your help

**Comment by fypeu** [comment]
Any solutions for this problem as of now?

**Comment by david-drinn** [comment]
Putting some thoughts down as I reason this out and try to figure out what the hold up is.

Taking the `Font` object as an example for discussion. It's an important part of a Style object.

### Let's Begin

Partly it's an issue of the fact that the Style/Font/etc. objects are _proxy objects_ rather than standalone objects. The underlying object is actually a node in the underlying XML tree of the document object. Styles and fonts can't exist on their own without the document it's attached to.

It's almost like we need a generic `Font` object, that is a standalone definition of an abstract font, outside of any specific document type. And then a converter of that to an XML node in a Word-specific doc. I'm guessing that's what a project like `pandoc` does, but I digress.

There is already an `add_style()` function as mentioned, but you can only pass in a name and type and "builtin" boolean flag. None of the meaningful data in the style, such as the font details and paragraph format, can be passed in. There's not even a proxy wrapper for other meaningful data about the style, such as Tabs, Border, Language, etc, but lets assume those aren't needed 

So you would need to manually iterate and copy properties from the following:

1. The `style` object itself, for things such as `style.hidden`, `style.locked`, `self.unhide_when_used` etc.
2. The `style.font` object, for things such as `style.font.all_caps`, `style.font.bold`, `style.font.cs_bold`, etc.
3. The `style.font.color` object, for `style.font.color.rgb` and `style.font.color.theme_color`. (It appears to be just those two.)
4. The `style.paragraph_format` object, for things such as `style.paragraph_format.alignment`, `style.paragraph_format.first_line_indent`, `style.paragraph_format.keep_with_next`, etc.
5. The `style.paragraph_format.tab_stops` object, for things such as `tab_stops.alignment`, `tab_stops.leader`, and `tab-stops.position`.  (It appears just those three.)
6. Possibly more that I missed.

### But wait

An important wrinkle here, is that most styles are based on other styles, in a cascading style tree. For example, the `Heading 1` style is based on the `Normal` style. This is important because, you can't know the true end result of all the style formatting options _without looking at all the styles_ it's based on.

For example, your style may not have set the Font's Bold option, ie. it shows as `None` -- so you might think it's for sure not bold, right? But oh wait, your style's _base style_ set it to True, or the next base style did, or the next... So it actually is Bold, _even though the style says it's not_! The common cascading style problem.

This cascading style makes sense I suppose for end users using the actual Word interface. You would like a user to only have to set a Font change once at a base level and have it cascade to all the other styles. And it also reduces file size and memory  and (possibly) processing requirements, which were more important 20-30 years ago. Personally, I think this has out-lived its welcome. Just... have each style completely live on its own, and then have a plethora of functions to do common user needs, such as updating all the fonts in the document at once to a new font.

Anyway, you have to traverse it. You _don't have to keep doing it, though_, not with the new doc. After traversing the tree, you could just add styles as completely standalone styles with no base styles. Whether you do that or not depends on your end use case. Power users of Word probably won't like it and would revert it. Average Word users, just adding content and making minimal style changes like bolding or italicizing here and there, probably wouldn't even notice the difference. Then there's APIs for tools that help generate Word docs and heavily rely on Styles that may or may not like it. 

Then there's number list styles. The meaningful formatting of these styles aren't supported at all currently in `python-docx`. So there's no meaningful way to copy those styles except in name and type only. That may be acceptable though.

### Summary

In summary, I think it's doable to add an `add_style(style)` signature, with perhaps a switch to make it recursively create all the base styles or just make one standalone style. A list of properties of each object would need to be maintained that should be copied over to the new object.

### P.S.

I think it would be good to have alternate properties (or switches to the existing properties) that tell you the true final value of things such as `font.bold`, without making the user have to cascade through everything to figure it out.

**Comment by scanny** [comment]
Yeah. Copying API objects (proxy objects) is definitely not going to work.

Theoretically, copying the style XML object the proxy object _wraps_ _could_ work but won't in the general case and probably not even most of the time.

The thing is, as @david-drinn rightly observes, that a style is not _atomic_. It involves _references_, or relationships to things in the source document that may or may not exist in the target document. There's also the case where the thing referenced by name in the source document _does_ exist in the target document, but behaves differently or has different settings such that the new style _works_, but _doesn't work (look) the same_ as it does in the source document.

I think the approach I would take to explore further is to write some code that discovers the attributes of a style in the source document then creates a style with those characteristics in the target document.

Then use that to discover all the nitty-gritty edge cases that make this a hard problem to solve in the general case.

Also, it could be the _actual_ case you need to solve is much simpler than the general case and an approach like this will do the trick for you.
